### PR TITLE
Al/pr support

### PR DESCRIPTION
**Pull request by dkjellin**
This is a big PR, it is mostly Amanda's work.

### CODE REVIEW

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (ysun-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:133]: The comments say that it should support pr. Does this mean that all the pull request events should be added here?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (ysun-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:216]: Why does it only check against `PULL_REQUEST_OPENED`? I may be misunderstanding, but shouldn't it check whether `events` contains any pull request event?

**Review by ysun-atl — CHANGES_REQUESTED** [review]

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:140]: /nit 
```suggestion
                supportedEvents.addAll(ALL_PULL_REQUEST_EVENTS);
```
Or do away with the list entirely and explicitly list supported events in the `process` method (I don't mind either way, but we should be consistent)

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:23]:
```suggestion
 * 1. Separate webhooks will be added for repo ref, pull request events (opened, updated and closed) and mirror sync events
```
or similar?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:37]: This is only being used once, so it can probably be inlined in the `process` method

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:40]: /nit
```suggestion
    private static final Set<String> REF_PR_EVENT_IDS = new HashSet<>();
```

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:169]: I think we really need to consider refactoring this method before we merge the PR. The existing system of just checking if we have matching triggers (which worked when there were only 2) is really straining now that we have 7, with four separate webhook lists we have to compare against. The lengthy conditional chains especially feel like an easy place for bugs to hide, and make detrimental changes accidentally.
I think this is complex enough to merit an object representation, and exposing methods like `hasWebhook`, so all these comparisons can be done internally. It would also mean we can add unit testing.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:216]: This is checking if we should perform a build, which we do when a push is made to a repo or a PR is opened. I notice that this conditional _doesn't_ check for source branch updated events (presumably because they're covered by repo ref change), but I think this would be more correct:
 ```suggestion
        if (events.stream().anyMatch(event -> refAndPREventIds.contains(event.getEventId()))) {
```

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:203]: Same as below, this isn't checking source branch updated (though I don't know if that would actually lead to an issue)

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/MinimalPullRequest.java
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/MinimalPullRequest.java:48]: and add it to the `hashCode` method too

**Review by mhenschke-atl — CHANGES_REQUESTED** [review]
Some nits but my big concern is the BitbucketWebhookConsumer- I think we should look at a different approach to process webhooks before we merge.
Will cover tests in a separate pass.

**Comment by driverpt** [comment]
Is there any ETA on this?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/MinimalPullRequest.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/MinimalPullRequest.java:12]: This class is not used, we should probably remove it

File: src/main/resources/com/atlassian/bitbucket/jenkins/internal/scm/SelectBranchTrait/help.groovy
Comment (dkjellin) [inline src/main/resources/com/atlassian/bitbucket/jenkins/internal/scm/SelectBranchTrait/help.groovy:1]: without the trait we don't need this help file

File: src/main/resources/com/atlassian/bitbucket/jenkins/internal/scm/SelectBranchTrait/help.properties
Comment (dkjellin) [inline src/main/resources/com/atlassian/bitbucket/jenkins/internal/scm/SelectBranchTrait/help.properties:1]: without the trait we don't need this translation

File: src/main/resources/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookTriggerImpl/help-pullRequestTrigger.html
Comment (dkjellin) [inline src/main/resources/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookTriggerImpl/help-pullRequestTrigger.html:2]: or source branch is updated

File: src/main/resources/com/atlassian/bitbucket/jenkins/internal/trigger/Messages.properties
Comment (dkjellin) [inline src/main/resources/com/atlassian/bitbucket/jenkins/internal/trigger/Messages.properties:4]: @mhenschke-atl are we removing the `Server` part of the name?

Comment (mhenschke-atl) [reply src/main/resources/com/atlassian/bitbucket/jenkins/internal/trigger/Messages.properties:4]: No word on that yet, but I'll check.

**Comment by dkjellin** [comment]
> Is there any ETA on this?

@driverpt Unfortunately no. The team working on this had to drop this to work on some more urgent things. That work is winding down but is not yet complete. We are fully aware of how important this is and are trying very hard to get this shipped as soon as possible. Thank you for your patience.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookTriggerImpl.java
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookTriggerImpl.java:249]: Sounds good.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookTriggerImpl.java
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookTriggerImpl.java:85]: Not quite sure what you're referring to here?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/CustomGitSCMSource.java
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/CustomGitSCMSource.java:20]: Yeah this should be removed altogether I think.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookTriggerImpl.java
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookTriggerImpl.java:85]: We have `if (version < BUILD_ON_PULL_REQUEST_VERSION)` and `f (version >= BUILD_ON_PULL_REQUEST_VERSION) {` they are the same in terms of the logic they are enforcing, but we should decide on one way to do this and stick with it. Personally I like the first one better `if(version<LATEST_VERSION) { return default value` but it is a weak preference.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketProject.java
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/model/BitbucketProject.java:47]: It should be final too (which is a bit of a nuisance, so I'll keep this comment open for reviewing.)

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRepositoryClient.java
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRepositoryClient.java:48]: Yeah, we can do that, it maps to the `ALL` query we pass to Bitbucket.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java:275]: I was about to, but realized I don't know why. It looks the "trigger anyway" clause, which we do if the job _isn't_ a computed folder (I guess a different job type we don't support?), or the event is null, but I couldn't really tell from the documentation what no event means. Or is this all just a way of saying we trigger "if we can", i.e. the consumer has supplied us with the information we need to trigger?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRepositoryClient.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRepositoryClient.java:45]: I think the documentation here is wrong too- it's a stream of all pull requests of the provided state right? (Open, merged, declined etc.)

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookTriggerImpl.java
Comment (mhenschke-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookTriggerImpl.java:85]: Oh I see the two different trigger classes. Will fix.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRepositoryClientImpl.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRepositoryClientImpl.java:39]: Maybe make a constant, or add to `BitbucketPullRequestState`?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookMultibranchTrigger.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookMultibranchTrigger.java:36]: I'm still a bit naff about this, but it's not a PR blocker. We should revisit if we add any more triggers in the future.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookMultibranchTrigger.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookMultibranchTrigger.java:61]: To make this just a hair more explicit (and extensible), maybe this?
`else if (event instanceof RefsChangedWebhookEvent) {
    return isRefTrigger();
}
return false; // No other events are applicable for triggers`

**Review by mhenschke-atl — APPROVED** [review]
I'm happy for this to merge :grinning:

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/scm/BitbucketSCMSource.java:279]: my memory is hazy, I _think_ the following holds
1. when a user clicks the 'scan now' button, no event is associated with that call so it is null here.
2. From memory there are a few calls along the way where event is null for perfectly fine reasons

Both of those should trigger a scan

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookMultibranchTrigger.java
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/BitbucketWebhookMultibranchTrigger.java:36]: I agree, but seeing as this is the DataboundConstructor we must have a field for every option in the UI, so.. at least here I think we need to have the listing.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:36]: These fields are actually not used at all any more, I've removed them all

**Review by ysun-atl — APPROVED** [review]

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (KalleOlaviNiemitalo) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:173]: Should this be `update(serverSideWebhooks, request, desiredEvents)` instead? I have not tested this, but from reading the code, it seems that `webhooks` contains all the webhooks that Bitbucket Server has on this repository, and the `update` call will then delete the webhooks that were not set by Bitbucket Server Integration plugin.

**Comment by KalleOlaviNiemitalo** [comment]
It seems that this does not create Jenkins jobs for the pull requests themselves but rather for their source branches. Is that correct?

Evidence:

- The `BitbucketRepositoryClient#getPullRequests` methods are called by `BitbucketRepositoryClientImplTest` only, i.e. they are not called when a user clicks "Scan Multibranch Pipeline Now".
- The plugin does not define a class that implements `ChangeRequestSCMHead2`; instead, `BitbucketWebhookConsumer.BitbucketSCMHeadPullRequestEvent` just uses `GitBranchSCMHead`.  In contrast, the [GitHub Branch Source](https://github.com/jenkinsci/github-branch-source-plugin/blob/d07e7d20a8458a2a1288ddbb85134abe30805119/src/main/java/org/jenkinsci/plugins/github_branch_source/PullRequestSCMHead.java#L48), [Bitbucket Branch Source](https://github.com/jenkinsci/bitbucket-branch-source-plugin/blob/c79f7c340cac7455b90d963bdaba9f4fb7dd2c25/src/main/java/com/cloudbees/jenkins/plugins/bitbucket/PullRequestSCMHead.java#L48), [Gitea](https://github.com/jenkinsci/gitea-plugin/blob/62855b3c3d8c119302b1156f5d3b0b3460810f76/src/main/java/org/jenkinsci/plugin/gitea/PullRequestSCMHead.java#L32), and [P4](https://github.com/jenkinsci/p4-plugin/blob/4a564282febb6a94a5833c02c0fce1f3d7e657cc/src/main/java/org/jenkinsci/plugins/p4/scm/P4ChangeRequestSCMHead.java#L12) plugins define such classes.

It would be useful to have Jenkins job for the pull request itself and let other plugins query information about it. In particular, [Branch API](https://github.com/jenkinsci/branch-api-plugin/blob/0f9cf838b6298d21652400dde9d611b3374f5b60/src/main/java/jenkins/branch/BranchNameContributor.java#L70-L71) would call [ChangeRequestSCMHead#getTarget()](https://github.com/jenkinsci/scm-api-plugin/blob/400faa8fd5651731257ff59f53b715dfeede7ede/src/main/java/jenkins/scm/api/mixin/ChangeRequestSCMHead.java#L50-L57) and set the `CHANGE_TARGET` global variable, which could then be passed to the [discoverGitReferenceBuild](https://www.jenkins.io/doc/pipeline/steps/git-forensics/#discovergitreferencebuild-discover-git-reference-build) pipeline step in order to let [Warnings Next Generation](https://plugins.jenkins.io/warnings-ng/) compare the build warnings of the pull request to those of its target branch.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:173]: Hello @KalleOlaviNiemitalo

I will look at this to ensure we're not deleting any webhooks that are not "ours". Thank you for reporting this.

**Comment by dkjellin** [comment]
>t seems that this does not create Jenkins jobs for the pull requests themselves but rather for their source branches. Is that correct?

That is correct, we are aware that this does not satisfy all usecases, but it is a stepping stone to supporting more usecases around pull requests.

Providing more information about the Pull request is something we can absolutely do, it will not make the `3.0.0` release as that is already long overdue, but if you open a ticket we can put it in a future release.

Thank you for your interest and attention to detail.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/trigger/register/BitbucketWebhookHandler.java:173]: It was a bug, I have put up a fix for it in https://github.com/jenkinsci/atlassian-bitbucket-server-integration-plugin/pull/273/files
thank you for reporting it!

**Comment by KalleOlaviNiemitalo** [comment]
> if you open a ticket we can put it in a future release.

OK, I filed [JENKINS-66581](https://issues.jenkins.io/browse/JENKINS-66581) for that.

**Comment by dkjellin** [comment]
> > if you open a ticket we can put it in a future release.
> 
> OK, I filed [JENKINS-66581](https://issues.jenkins.io/browse/JENKINS-66581) for that.

Thank you for the detailed ticket. Once we (finally) get this released we can look at maybe resolving that ticket.

**Pull request merged**
